Feature/test for optimal path task

YuriyMasechko opened the pull request:
New test for optimal path task was added.
Test has three scenarios:
- test gives 11k response that shows best link quality for current BSS, no steering done
- test gives 11k response that shows best link quality in a different BSS, steer client
- error response to 11k request, optimal path falls back to RSSI measurement

Additionally dhcp events were added as association task
requires sta to be connected with IP. New data path to change sta stats was added as event.

fixes #443

YuriyMasechko committed:
bwl: separate dummy events to EVENT and DATA

It is required to update data via dummy events. For example update_stations_stats
is required to restart optimal path task. To better understanding
what is happening need to split dummy events and dummy data
notifications.

Separate dummy events to EVENT and DATA.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
bwl: add dummy beacon 11k request

For test purpose need to add dummy beacon 11k request.

Add dummy beacon 11k request.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
bwl: add dummy RRM_Beacon_Response support

For test purpose need to add beacone 11k response.

Add dummy RRM_Beacon_Response support.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
bwl: add STA_Update_Stats event

To restart optimal path task need to update statio stats. New dummy
event need to be added.

Add STA_Update_Stats event.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

morantr commented on framework/platform/bpl/linux/bpl_dhcp_mon.c:
```diff
@@ -8,8 +8,74 @@
+    // Execute the callback
+    if (s_pCallback)
+        s_pCallback((arrTok[0]) ? arrTok[0] : "", (arrTok[1]) ? arrTok[1] : "",
+                    (arrTok[2]) ? arrTok[2] : "", (arrTok[3]) ? arrTok[3] : "");
```
I think it is a good idea to add a description of the tokens:
<op: {"add","del", "old"}> <client_mac> <ip> <hostname>

Also, create some data validation. But IMO it should be on platform_manager_thread.cpp:send_dhcp_notification().
since it is more convenient to validate there the mac, the IP and the op string.

YuriyMasechko replied:
Comment was changed.
Open separate task to add data validation to manager?

morantr commented on common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp:
```diff
@@ -120,9 +120,14 @@ static void map_event_obj_parser(std::string event_str, parsed_obj_map_t &map_ob
+            map_obj[DUMMY_EVENT_KEYLESS_PARAM_TYPE] = str_storage;
+            type                                    = false;
+        } else if (opcode) {
+            // assume that the second param is name
```
please change the comment -> "....is data or event name"

YuriyMasechko replied:
Done.

morantr commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:
```diff
@@ -75,7 +83,13 @@ bool mon_wlan_hal_dummy::update_vap_stats(const std::string vap_iface_name, SVap
 bool mon_wlan_hal_dummy::update_stations_stats(const std::string vap_iface_name,
```
If you can, please change the string argument to be const reference.

YuriyMasechko replied:
It will require to change all bwl implementations, maybe separate task?

morantr replied:
So leave it for now...

morantr commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.h:
```diff
@@ -54,6 +62,7 @@ class mon_wlan_hal_dummy : public base_wlan_hal_dummy, public mon_wlan_hal {
     // Private data-members:
 private:
     std::shared_ptr<char> m_temp_dummy_value;
+    SStaStats m_dummy_sta_stats;
```
This implementation supports only one station.
I think it will be better if you store it on unrodered_map which it's key is "sta_mac".
Optional: create a handler for a client disconnect event to remove the client element from the map.

tomereli replied:
Good idea, but if it is too much work than I suggest to open a task for it. We won't need dummy implementation to test more than one dummy station for now.

YuriyMasechko replied:
Extended to map.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+
+    dbg "DHCP event"
+    check docker exec -it gateway sh -c \
+        'echo "0,11:22:33:44:55:66,192.168.1.10,test" > /tmp/$USER/beerocks/dhcp/EVENT'
```
What is the zero at the string beginning? shouldn't it be one of {"add","del","old"} ?
Please look for main_thread::send_dhcp_notification on platform manager.

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+    check docker exec -it repeater1 sh -c \
+        "grep -i -q 'Beacon 11k request 11:22:33:44:55:66 aa:bb:cc:00:00:10 1' /tmp/$USER/beerocks/logs/beerocks_monitor_wlan0.log"
+
+    dbg "RPM Beacon Response"
```
I think you meant "RRM"? If not, what is "RPM" then?

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+    check docker exec -it repeater1 sh -c \
+        'echo "EVENT RRM-BEACON-REP-RECEIVED 11:22:33:44:55:66 channel=1 dialog_token=0 measurement_rep_mode=0 op_class=0 duration=50 rcpi=-40 rsni=40 bssid=aa:bb:cc:00:00:10" > /tmp/$USER/beerocks/wlan0/EVENT'
+
+    dbg "RPM Update Stats"
```
If indeed "RRM" is what you meant, so it is defintly not related to regular measurement, so please remove it from the comment.

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+    check docker exec -it repeater1 sh -c \
+        "grep -i -q 'Beacon 11k request 11:22:33:44:55:66 aa:bb:cc:11:00:20 149' /tmp/$USER/beerocks/logs/beerocks_monitor_wlan0.log"
+
+    dbg "RPM Beacon Response"
```
RRM?

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review

morantr commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:
```diff
@@ -80,6 +80,49 @@ bool mon_wlan_hal_dummy::sta_channel_load_11k_request(const SStaChannelLoadReque
+        cmd += " ssid=" + req_ssid;
+    }
+
+    LOG(DEBUG) << "Beacon 11k request " << cmd;
```
What is the purpose of the request struct processing? I mean, we don't send it to hostapd so it makes it makes the hostapd string construction to redundant.
The print only is enough: "Beacon measurement request to sta <sta_mac> on bssid <bssid_mac>" for the test_flow.sh.

tomereli replied:
Fully agree with Moran on this one.

YuriyMasechko replied:
Changed to print only mac, bssid and channel.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+    dbg "RPM Update Stats"
+    check docker exec -it repeater1 sh -c \
+        'echo "DATA STA-UPDATE-STATS rssi=-38,-39,-40,-41 snr=38,39,40,41 uplink=1000 downlink=800" > /tmp/$USER/beerocks/wlan0/EVENT'
+    sleep 15
```
Why sleeping 15 seconds? Isn't 1-2 seconds is enough?

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+    check docker exec -it gateway sh -c \
+        'echo "0,11:22:33:44:55:66,192.168.1.10,test" > /tmp/$USER/beerocks/dhcp/EVENT'
+
+    dbg "Confirming 11k request is done"
```
To make it more clear, I think that it will help to mention as part of which procedure/task this is done.
For example, this request is done by the client association task (right?).
And of course, apply it on the others as well.
To be honest, I find it very difficult to review it, since I see in this file so many events of requests, responses, etc..
that I'm not able to understand what exactly is going on the controller. I would appreciate comments that explain what is happening as getting forward with this test.

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+
+    dbg "RPM Beacon Response"
+    check docker exec -it repeater1 sh -c \
+        'echo "EVENT RRM-BEACON-REP-RECEIVED 11:22:33:44:55:66 channel=149 dialog_token=0 measurement_rep_mode=0 op_class=0 duration=50 rcpi=-80 rsni=10 bssid=aa:bb:cc:11:00:20" > /tmp/$USER/beerocks/wlan0/EVENT'
```
I don't understand how this script is synced with the controller. I mean, you can't know when exactly the request is sent, so how can't you know when to send an event on time so you will not reach timeout.
I think that the right thing to do is to store on the beacon measurement responses on the dummy for each bssid by sending it as DATA, and when receiving a measurement request, pull the data and simulate an internal "RRM-BEACON-REP-RECEIVED" event.
You can use the unordered map which I suggested on other comment to hold the sta measurement, to hold the beacon measurement response as well. 
That way you can pre-prepare the responses and make sure it will be sent on time.

```cpp
struct sDummySta {
    SStaStats sta_stats;
    std::unordered_map<std::string, std::string> beacon_measurment_events; // key=bssid, value=event string
};
std::unordered_map<std::string, sDummySta> dummy_sta; // key=sta_mac
```

YuriyMasechko replied:
New DATA event was added to store beacon measurement.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
         
 }
 
+test_optimal_path_dummy() {
```
hmm, does this test should be here? I thought we test here only certification flows.
@arnout

YuriyMasechko replied:
@arnout should it go to separate file?

morantr replied:
@tomereli can you please reply

arnout replied:
It should be a separate file (and the client steering test as well). However, that can be done in a follow-up PR. Putting it in a separate file will require a bit of refactoring, e.g. moving some functions to functions.sh, and maybe not doing the startup from this script but instead do it in a separate step. And it requires an update of travis.yml.

So yes, it should be a separate file, but that should be done in a new PR. @YuriyMasechko if you have time you could work on that.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+
+    dbg "RPM Beacon Response"
+    check docker exec -it repeater1 sh -c \
+        'echo "EVENT RRM-BEACON-REP-RECEIVED 11:22:33:44:55:66 channel=149 dialog_token=0 measurement_rep_mode=0 op_class=0 duration=50 rcpi=-80 rsni=10 bssid=00:11:22:33:00:20" > /tmp/$USER/beerocks/wlan0/EVENT'
```
You already requested measurement one time for each bssid (I assume it is part of optimal path task), why did you 2 more measurement on wlan0 and 1 more measurement on wlan2?
Same thing on the second optimal task phase.

YuriyMasechko replied:
Because there are 2 IREs.

morantr commented on tools/docker/tests/test_flows.sh:
```diff
@@ -207,6 +208,220 @@ test_client_association_dummy(){
+
+    dbg "Confirm steering success by client connected"
+    check docker exec -it gateway sh -c \
+        'grep -i -q "steering successful for sta 11:22:33:44:55:66" /tmp/$USER/beerocks/logs/beerocks_controller.log'
```
If I'm not mistaken, these tests are already done by other test flows, so isn't it redundant to do it again?
I think that it is enough to check that optimal_path has decided to steer the client, without checking the whole steering flow.
Another option is to use the functions that already test some flows (like steering) and combine them in the current test (as a function call I mean).

YuriyMasechko replied:
Will revert this test and commit new one. To big diff to review.

morantr reviewed (changes requested):
Besides my comments, please update the dummy bwl readme.md file.

tomereli commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.h:
```diff
@@ -20,6 +20,14 @@ namespace dummy {
+public:
+    enum class Data {
+        Invalid = 0,
+
```
redundant line breaks

YuriyMasechko replied:
Done.

YuriyMasechko committed:
bwl: add RRM_Update_Beacon_Measurements event

To have possibility to respond immediately after beacon measurement
request was received, need to pre-prepare beacon measurement responses
for all bssids.

Add RRM_Update_Beacon_Measurements event.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
tools: docker: add dummy test for Optimal Task

Add dummy test for Optimal Task.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

arnout commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:
```diff
@@ -52,6 +52,8 @@ static mon_wlan_hal_dummy::Data dummy_to_bwl_data(const std::string &opcode)
 {
     if (opcode == "STA-UPDATE-STATS")
         return mon_wlan_hal_dummy::Data::STA_Update_Stats;
+    else if (opcode == "RRM-BEACON-REP-RECEIVED")
+        return mon_wlan_hal_dummy::Data::RRM_Update_Beacon_Measurements;
```
We end up using only this DATA. So I think it's not actually needed to make this split into DATA/EVENT. Instead, we can put everything in process_dummy_event().

YuriyMasechko replied:
Data path is necessary for 2 reasons:
1. to prevent sleeping and un-synced operation on the test_flows.sh.
2. It is a preperation for the future, for commands that needs to return data immediatly that is not base on events, and we want the ability to change it.

Of course we could handle everything in process_event(), but it will be not such clear.

arnout commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:
```diff
@@ -80,6 +114,27 @@ bool mon_wlan_hal_dummy::sta_channel_load_11k_request(const SStaChannelLoadReque
+            //TODO: remove sleep. Sleep fixes missing requests sent by controller but not
+            //received by agent
+            sleep(1);
+            process_dummy_event(dummy_bssid_event->second);
```
This is the only place where the RRM event is used. So instead, that code can be moved here directly.

YuriyMasechko replied:
See comment above.

arnout commented on common/beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp:
```diff
@@ -95,15 +150,198 @@ bool mon_wlan_hal_dummy::sta_link_measurements_11k_request(const std::string &st
+    case Event::RRM_Beacon_Response: {
+        // Allocate response object
+        auto resp_buff = ALLOC_SMART_BUFFER(sizeof(SBeaconResponse11k));
+        auto resp      = reinterpret_cast<SBeaconResponse11k *>(resp_buff.get());
```
So this code can move directly into sta_beacon_11k_request().

YuriyMasechko replied:
See comment above.

arnout reviewed (changes requested):
I rebased on master again. I also removed the commit adding the dummy dhcp, since it's no longer needed (requirement for IP address has been removed).

arnout reviewed (approved):
Since my remaining comment was overridden, I'll approve and merge after CI finishes.
